src/java/main/org/apache/zookeeper/server/ZooKeeperServerMain.java

```diff
@@ -146,14 +146,26 @@ public void runFromConfig(ServerConfig config) throws IOException, AdminServerEx
             );
             containerManager.start();
 
+            // Watch status of ZooKeeper server. If there is an internal error
+            // then will do a graceful shutdown.
+            while (zkServer.isRunning()) {
```

Comment:
OK. How about exposing an interface like below by ZooKeeperServer.java class and ZooKeeperServerMain will implement this using CDL logic. This implementation will be passed as constructor arg in ZooKeeperServer.java and invoked during state changes.

/**
 * Observer interface which will be used to notify the state changes.
 */
public interface ZooKeeperServerStateListener {

    /**
     * This will be invoked when the server transition to a new server state.
     *
     * @param state new server state
     */
    void stateChanged(State state);
}

ZooKeeperServerMain.java
    private static class ZooKeeperServerStateListenerImpl
            implements ZooKeeperServerStateListener {
        private CountDownLatch healthMonitorLatch;

        ZooKeeperServerStateListenerImpl(CountDownLatch healthMonitorLatch) {
            this.healthMonitorLatch = healthMonitorLatch;
        }

        @Override
        public void stateChanged(State state) {
            if (state != State.RUNNING) {
                healthMonitorLatch.countDown();
            }
        }
    }